packages/commons/src/interop-token/models.ts

```diff
@@ -90,54 +113,105 @@ export type InteropApiToken = {
   payload: InteropJwtInternalPayload;
   serialized: string;
 };
 
-/* ========================================== 
-    Interop SESSION Token 
-  ========================================== */
-const Organization = z.object({
-  id: z.string(),
-  name: z.string(),
-  roles: z.array(z.object({ role: z.string() })),
-});
+// ==========================================
+//    Interop UI Token
+// ==========================================
 export const SessionClaims = z.object({
-  [UID]: z.string(),
-  [ORGANIZATION]: Organization,
-  [NAME]: z.string().optional(),
-  [FAMILY_NAME]: z.string().optional(),
-  [EMAIL]: z.string().optional(),
+  uid: UserId,
+  organization: z.object({
+    id: SelfcareId,
+    name: z.string(),
+    roles: z.array(
+      z.object({
+        partyRole: z.string().nullish(),
+        role: UserRole,
+      })
+    ),
+    fiscal_code: z.string().nullish(),
+    ipaCode: z.string().nullish(),
+  }),
+  name: z.string().nullish(),
+  family_name: z.string().nullish(),
+  email: z.string().nullish(),
 });
 export type SessionClaims = z.infer<typeof SessionClaims>;
 
-export const CustomClaims = z.object({
-  [USER_ROLES]: z.string(),
-  [ORGANIZATION_ID_CLAIM]: z.string(),
-  [SELFCARE_ID_CLAIM]: z.string(),
-  [ORGANIZATION_EXTERNAL_ID_CLAIM]: z.object({
-    [ORGANIZATION_EXTERNAL_ID_ORIGIN_CLAIM]: z.string(),
-    [ORGANIZATION_EXTERNAL_ID_VALUE_CLAIM]: z.string(),
+export const UserClaims = z.object({
+  "user-roles": CommaSeparatedStringToArray(UserRole),
+  organizationId: TenantId,
+  selfcareId: SelfcareId,
+  externalId: z.object({
+    origin: z.string(),
+    value: z.string(),
   }),
 });
-export type CustomClaims = z.infer<typeof CustomClaims>;
+export type UserClaims = z.infer<typeof UserClaims>;
 
-export type SessionJwtPayload = InteropJwtCommonPayload &
-  SessionClaims &
-  CustomClaims;
+export const UIClaims = SessionClaims.merge(UserClaims);
 
-export type SessionToken = {
+export type UIClaims = z.infer<typeof UIClaims>;
+
+export const InteropJwtUIPayload = InteropJwtCommonPayload.merge(
+  UIClaims
+).extend({
+  // setting role to z.undefined() to make the discriminated union work.
+  // z.discriminatedUnion performs better than z.union and gives more meaningful parsing errors.
+  role: z.undefined(),
+});
+
+export type InteropJwtUIPayload = z.infer<typeof InteropJwtUIPayload>;
+
+export type InteropUIToken = {
   header: InteropJwtHeader;
-  payload: SessionJwtPayload;
+  payload: InteropJwtUIPayload;
   serialized: string;
 };
+
+// ===========================================
+//    Parsing utilities
+// ===========================================
+
+// AuthTokenPayload is a discriminated union used to parse the payload of
+// the auth token we receive in API requests phase. It includes only the payloads
+// that we actually can receive. For example, it does not include the
+// InteropJwtConsumerPayload, because interop generates it but nevere receives it in API requests.
```

Comment:
```suggestion
// AuthTokenPayload is a discriminated union used to parse the payload of
// the auth token we receive in API requests. It includes only the payloads
// that we actually can receive. For example, it does not include the
// InteropJwtConsumerPayload, because interop generates it but never receives it in API requests.
```